### PR TITLE
Support for Git

### LINKED ISSUE #156: TM1 Git integration
I was going to start building out a GitService to incorporate the new features available in 2.07. Before I get started I wanted to make sure I wasn't duplicating work that was already underway.

Cheers.

### PR DESCRIPTION
**Pull request by lapstue**
This set of commits adds basic support for Git by adding a GitService and objects for Git, GitCommit, GitPlan and GitRemote. The code is based on a postman collection set I created to experiment with GIT and TM1 and has been working in my rather limited internal testing. 

I've tried to adhere to what I perceive to be the current software design patterns of tm1py. Feedback is welcomed, and please inform me if you want me to squash the commits or do other things to improve the merge. 

Fixes #156

### CODE REVIEW

**Comment by MariusWirtz** [comment]
Hi @lapstue,

At first glance, the code looks good and the design is all in line with the established design patterns.

I will try to test and review the code in depth this week and provide feedback.
Yes. It would be nice if you could squash the commits before we merge them.

Thank you for your contribution. Very much appreciated!

**Comment by lapstue** [comment]
@MariusWirtz Great to hear! 

I've done some testing myself, and it seems to be working fine, but getting someone else to look over the code and do their own testing and feedback is always a good idea. Fair warning, I haven't been able to test SSH support as the TM1 installs I have access to are older than 2.0.9.4, which added support for SSH, but it should work as intended based on the IBM documentation.

I'll hold off on squashing until I get some feedback.

Once this PR clears, I want to start looking more into the TM1Project API, as it's an important part of deploy using GIT. I looked into it a bit earlier, but ran into issues due to poor documentation.

**Comment by rclapp** [comment]
Awesome, looks like I can stash my old for then.

**Comment by lapstue** [comment]
@rclapp I wasn't aware of your fork, it would've probably been easier to pick up from where you stopped. Nevertheless, in the end we ended up with very similar implementations. Please do point out if there's something in your fork I should include in the pull request.

**Comment by rclapp** [comment]
Sounds good 😊

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [inline TM1py/Services/GitService.py:126]: To be pep-8 compliant this should be called `git_execute_plan` and `plan_id` :)

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [inline TM1py/Services/GitService.py:131]: You can use the `format_url` function if there are characters in the `plan_id` that must be url-escaped

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [inline TM1py/Services/GitService.py:134]: `git_get_plans`

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [inline TM1py/Services/GitService.py:144]: To reduce indentation by one level you can write

``` python
if plan_id is None:
    continue
```

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [inline TM1py/Services/GitService.py:168]: I think it would be good to use `elif` here and ideally include an `else` clause that throws a `ValueError` indicating that the plan is illegal

**Review by MariusWirtz — CHANGES_REQUESTED** [review]
I ran the code on my end. So far it looks like everything is working as it should.

I requested a few changes mostly related to style.
What do you think of my remarks w.r.t. the `kwargs` arguments?

Please also squash the commits.

**Comment by MariusWirtz** [comment]
In case anyone wants to test this new feature. Here is a simple implementation:


``` python
from TM1py import TM1Service

# Setup
# 1. create private repo in GitHub
# 2. follow instructions to add ssh key to GitHub profile
# https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent


with TM1Service(address="", port=12297, ssl=True, user="admin", password="apple") as tm1:
    # copy from GitHub repo
    url = 'git@github.com:MariusWirtz-cubewise/sdata.git'

    # can be anything
    deployment = 'prod'

    # as specified during creation of ssh key
    with open(r'C:\Users\MariusWirtz\.ssh\id_rsa', 'r') as file:
        private_key = file.read()
    with open(r'C:\Users\MariusWirtz\.ssh\id_rsa.pub', 'r') as file:
        public_key = file.read()
    passphrase = '*****'

    tm1.git.git_init(
        URL=url,
        Deployment=deployment,
        PublicKey=public_key,
        PrivateKey=private_key,
        Passphrase=passphrase)

    tm1.git.git_push(
        execute=True,
        Branch="feature-a",
        Message="Initial Commit",
        Author="Marius",
        Email="MWirtz@cubewise.com",
        PublicKey=public_key,
        PrivateKey=private_key,
        Passphrase=passphrase)

    plans = tm1.git.git_getplans()

    print(plans)
```

**Comment by lapstue** [comment]
@MariusWirtz 

> What do you think of my remarks w.r.t. the `kwargs` arguments?

As you pointed out, the upside of using kwargs is that it saves a lot of code and checking for empty/none function arguments + assigning variables when constructing the GitInit POST request. Another advantage of kwargs is that it gives some flexibility when/if the available parameters in GitInit change and TM1py hasn't been updated to reflect the new options. I'll still leave kwargs in the method definition to allow for custom headers and such, so that will address such cases. 

I'll go through the documentation again to make sure that I fully understand what REST API arguments are needed in all GitInit requests, and all that are optional/situational. 

What's the preferred way of checking if the arguments are unset? A bunch of

```
if arg1 is not None:
    _arg1 = arg1
```
or is there a better way of handling it? It quickly turns into a rather large block of checks to build the resultant dict. I guess we could add a utility like remove_none_value_arguments(**kwargs) that loops through the arguments, ignores the empty ones and returns a list of arguments with None values.

> Please also squash the commits.

I'll do this as I commit the changes you outlined above. Thanks for the concise feedback! I'll try to get it done in the next couple of days, if time permits. I have also started some work on a new TM1Project branch, and will try to incorporate the feedback into that branch as well, to reduce the feedback burden once I finish TM1Project and submit a pull request.

**Comment by lapstue** [comment]
@MariusWirtz 

Okay, I now believe that I've implemented all the changes you asked for, long story (not so) short:

- Kwargs has been replaced with named parameters for the proper REST API parameters, but **kwargs has been left in place to allow for custom headers and so on. Type hints have been added for the new named parameters. 
- Git-objects are no longer subclassed form TM1Object, as I found no proper way of doing it and adhere to internal standards. We might have to revisit this later on, but it will work for the time being. 
- Format_url is used instead of .format were applicable. 
- Pep8 should now be respected fully, unless my IDE is missing something.
- Private variables of objects are now prefixed with "_" and getters are decorated. 
- Proper import order is now implemented, to ensure that we don't get recursive imports
- After a bit of messing about the commits are now squashed into a nice single commit.
- 
So, round two of reviews I guess?

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [inline TM1py/Services/GitService.py:75]: This is not necessary. If you just pass the kwargs through to the POST method the timeout and other technical arguments will be respected.

``` python
self._rest.POST(url=url, data=body, **kwargs)
```

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [inline TM1py/Services/GitService.py:114]: same here :)

**Comment by MariusWirtz** [comment]
> @MariusWirtz
> 
> Okay, I now believe that I've implemented all the changes you asked for, long story (not so) short:
> 
> * Kwargs has been replaced with named parameters for the proper REST API parameters, but **kwargs has been left in place to allow for custom headers and so on. Type hints have been added for the new named parameters.
> * Git-objects are no longer subclassed form TM1Object, as I found no proper way of doing it and adhere to internal standards. We might have to revisit this later on, but it will work for the time being.
> * Format_url is used instead of .format were applicable.
> * Pep8 should now be respected fully, unless my IDE is missing something.
> * Private variables of objects are now prefixed with "_" and getters are decorated.
> * Proper import order is now implemented, to ensure that we don't get recursive imports
> * After a bit of messing about the commits are now squashed into a nice single commit.
> 
> So, round two of reviews I guess?

Thanks for the changes! It's looking really good. I will do more heavy testing and see if I can break anything :)

**Comment by lapstue** [comment]
The code has been changed and *kwargs is passed directly instead of being iterated and added to the body dict. 

Please let me know if you're able to pinpoint any non-working (corner) cases and I'll do my best to fix those. If you're unable to find any problems, I think we're pretty much done when it comes to basic git support :)

**Comment by MariusWirtz** [comment]
@lapstue,
It's looking all good. I will merge the changes after we finished the 1.6 release tomorrow. 
I would prefer to keep the git related changes in the master branch for a bit to mature and include them in the next major release in a couple of weeks.

Thanks again for the contribution. Much appreciated!

**Review by MariusWirtz — APPROVED** [review]

**Comment by MariusWirtz** [comment]
Fixes #156

**Comment by scrumthing** [comment]
Tried it out and it is awesome. Thanks for developing it @lapstue. I wonder if one should add the **GitDeploy** method. What do you think?

` <Action Name="GitDeploy">
<Annotation Term="Core.Description">
<String>Deploy a model using the source from the specified Git repository.</String>
</Annotation>`

What is your idea for the tm1project file? TM1py could assist with the creation of that file. What do you think?

**Comment by lapstue** [comment]
> Tried it out and it is awesome. Thanks for developing it @lapstue. I wonder if one should add the **GitDeploy** method. What do you think?

GitDeploy is pretty much the only piece missing from Git right now. I didn't implement it in my pull request because it isn't something I've experimented with so far. I'll try to look into it in the next couple of weeks.

> What is your idea for the tm1project file? TM1py could assist with the creation of that file. What do you think?

I want to add full support for TM1Project, with objects for ignore, tasks and deployments. TM1Project is a bit rough around the edges in the current IBM implementation (control objects can't be included using a wildcard for example), so combining it with the methods and logic exposed in TM1Py would be very useful.

File: TM1py/Services/GitService.py
Comment (rkvinoth) [inline TM1py/Services/GitService.py:60]: We can just loop through `locals()` here.

File: TM1py/Services/GitService.py
Comment (rkvinoth) [inline TM1py/Services/GitService.py:100]: Again... Loop through `locals()`.

File: TM1py/Services/GitService.py
Comment (rkvinoth) [inline TM1py/Services/GitService.py:137]: Again... Loop through `locals()`.

File: TM1py/Services/GitService.py
Comment (rkvinoth) [inline TM1py/Services/GitService.py:183]: Again... Loop through `locals()`.

**Review by rkvinoth — CHANGES_REQUESTED** [review]

File: TM1py/Services/GitService.py
Comment (rkvinoth) [inline TM1py/Services/GitService.py:116]: `force` and `execute` are of type `bool` then the default value should be `False`

File: TM1py/Services/GitService.py
Comment (rkvinoth) [inline TM1py/Services/GitService.py:170]: `force` is of type `bool` then the default value should be `False`

File: TM1py/Services/GitService.py
Comment (rkvinoth) [inline TM1py/Services/GitService.py:165]: `git_execute_plan` could fail silently here. We need some exception handling in place.

**Review by rkvinoth — CHANGES_REQUESTED** [review]

**Comment by scrumthing** [comment]
> GitDeploy is pretty much the only piece missing from Git right now. I didn't implement it in my pull request because it isn't something I've experimented with so far. I'll try to look into it in the next couple of weeks.

From my experience GitDeploy is merely a combination of GitInit and GitPull for setting up new databases. 

> I want to add full support for TM1Project, with objects for ignore, tasks and deployments. TM1Project is a bit rough around the edges in the current IBM implementation (control objects can't be included using a wildcard for example), so combining it with the methods and logic exposed in TM1Py would be very useful. Right now I'm maintaining TM1Project files with Postman and a manual edit process, which is quite tedious. I've been working on a GUI for TM1 + Git for internal use for the modellers with a less technical background, but I need to get TM1project sorted before that can happen. 

Yes. The tm1project file needs some tinkering with. I am not sure adding control objects to the file is the way to go. Had some issues with aliases because it seems they are not properly recognized and get deployed as strings. Picklist cubes with rules on the other hand are transferred nicely. Still not sure if I want to write processes for everything I put into control cubes but with git that seems to be the way. Could be annoying for security stuff. But as we all know rules are no proper solution in that case.

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:260]: Changed it to `RuntimeError`

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:264]: Agree

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:220]: Changed it to:

``` python
        response = self._rest.GET(url=url, **kwargs)
        for plan in response.json().get('value'):
```

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:100]: If you loop through locals, how do you map the variable name (e.g. username, private_key) to their expected keys in the JSON (e.g. Username, PrivateKey) ?

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:170]: I think it's kinda valid. If `force` is not passed to the function, it will not be included in the body of the request and TM1 would use the default.

Do you think the default behavior should be controlled in TM1py instead through a default value for the `force` argument?

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:165]: good catch

File: TM1py/Services/GitService.py
Comment (rkvinoth) [reply TM1py/Services/GitService.py:170]: IMO having `False` there will make the code look good but would leave it to you to decide.

File: TM1py/Services/GitService.py
Comment (rkvinoth) [reply TM1py/Services/GitService.py:260]: Sounds good

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:170]: OK

File: TM1py/Services/GitService.py
Comment (MariusWirtz) [reply TM1py/Services/GitService.py:165]: Are you sure about this one? In the happy case it returns a response with status code 204. 
If I tweak the code and pass a wrong `plan_id` to the function, it raises a TM1pyRestException.

In which case does it fail silently?

**Pull request merged**
